src/H5Tcommit.c

```diff
@@ -662,7 +663,7 @@ H5T__open_api_common(hid_t loc_id, const char *name, hid_t tapl_id, void **token
 done:
     /* Cleanup on error */
     if (H5I_INVALID_HID == ret_value)
-        if (dt && H5VL_datatype_close(*vol_obj_ptr, H5P_DATASET_XFER_DEFAULT, H5_REQUEST_NULL) < 0)
```

Comment:
This VOL object pointer actually points to the VOL object that was passed in to `H5Topen2` for `loc_id` rather than pointing to the VOL object for the committed datatype. When `H5VL_register` above fails, `dt` is left as a typeless pointer not attached to a VOL object, so we have to wrap it in one before being able to call `H5VL_datatype_close` on it.